lib/tooling/llvm-mca-tool.ts

```diff
@@ -47,21 +47,56 @@ export class LLVMMcaTool extends BaseTool {
 
     override async runTool(compilationInfo: Record<any, any>, inputFilepath?: string, args?: string[]) {
         const isets = new InstructionSets();
-        const target = isets.getInstructionSetTarget(compilationInfo.compiler.instructionSet);
-        if (target != null) args?.push('-mtriple=' + target);
+        let target = isets.getInstructionSetTarget(compilationInfo.compiler.instructionSet);
+        const prependArgs: string[] = [];
+        let haveCPU = false;
 
+        // Check if compiler target is overridden with --target.
         for (const arg of compilationInfo.options) {
-            if (arg.startsWith('-mcpu') || arg.startsWith('-march') || arg.startsWith('-mtriple')) {
-                args?.push(arg);
+            if (arg.startsWith('--target=')) {
+                target = arg.replace(/^--target=/, '');
+            }
+        }
+
+        if (target != null) prependArgs.push('-mtriple=' + target);
+
+        for (const arg of compilationInfo.options) {
+            if (arg.startsWith('-march')) {
+                if (target?.startsWith('x86_64')) {
+                    // On x86 in clang the CPU is specified with march,
+                    // but MCA requires -mcpu to function, so rewrite it.
+                    prependArgs.push(arg.replace(/^-march=(?!x86-64)/, '-mcpu='));
+                }
+            }
+            if (arg.startsWith('-mcpu') || arg.startsWith('-mtriple')) {
+                prependArgs.push(arg);
+            }
+            if (arg.startsWith('-mcpu')) {
+                haveCPU = true;
             }
         }
 
         if (compilationInfo.filters.binary || compilationInfo.filters.binaryObject) {
             return this.createErrorResponse('<cannot run analysis on binary>');
         }
 
+        // If there is a target specified which is not x86, and there is
+        // no other CPU specified, default to mcpu=generic to override
+        // llvm-mca's default of -mcpu=native. Otherwise MCA uses
+        // -mcpu=native for a target which is different from that
+        // compiler-explorer is running llvm-mca on, and it results in
+        // an error that MCA can't find the CPU for the target.
+        if (target && !target.startsWith('x86_64') && !haveCPU) {
+            prependArgs.push('-mcpu=generic');
```

Comment:
I think `generic` is also a reasonable default for x86_64 as well (and x86 for the matter, so always). Otherwise llvm-mca successfully finds native target, and then results depend on a server which processed this run.

Reply:
> A bit strange that compiler defaults to some generic target but llvm-mca defaults to native.

I take it that the compiler defaults to a generic CPU because this ensures wide compatibility of the compiled binaries (rather than CPU-specific), whereas MCA defaults to giving you a performance analysis of the machine the user is on, as a convenience. This makes reasonable sense to me but doesn't work if you're running MCA in a different environment than you are targetting, such as in on compiler-explorer.

I like the idea of using -mcpu=generic if mcpu (or on x86, march) is not otherwise specified on compiler-explorer.

Reply:
Done.